File: cracking_coding_interview/chap_4/4.1_route_between_nodes.js
const routeBetweenNodes = (startNode, searchNode) => {
  if (!startNode || searchNode) return false;
  if (startNode === searchNode) return true;

  const queue = [startNode];

  while (queue.length > 1) {
    const curNode = queue.shift();

    if (curNode === searchNode) return true;

    curNode.children.forEach(node => {
      if (!node.visited) queue.push(node);
      node.visited = true;
    });
  }
  return false;
};

File: cracking_coding_interview/chap_4/4.1_route_between_nodes.rb
def route_between_nodes(start_node, search_node)
  return false if !start_node || !search_node
  return true if start_node == search_node

  queue = [start_node]

  until queue.empty?
    cur_node = queue.shift

    return true if cur_node == search_node

    cur_node.children.each do |node|
      queue.push(node) if !node.visited
      node.visited = true
    end
  end
  false
end
